Update the microarray UMAP example notebook (#200)
AlexsLemonade/refinebio-examples: 02-microarray/dimension-reduction_microarray_02_umap.Rmd

# Check if this is in the same order
all.equal(colnames(df), metadata$geo_accession)
```

Now we are going to use a combination of functions from the `umap` and the `ggplot2` packages to perform and visualize the results of the Uniform Manifold Approximation (UMAP) dimension reduction technique on our medulloblastoma samples.

## Perform Uniform Manifold Approximation (UMAP)

In this code chunk, we are going to perform Uniform Manifold Approximation (UMAP) on our data and create a data frame using the UMAP scores and the variables from our metadata that we are going to use to annotate our plot later.

```{r}
# Perform Uniform Manifold Approximation (UMAP)
umap_data <- umap::umap(t(df)) # We have to transpose our data frame to meet the `umap::umap()` function requirements

# Make into data frame for plotting with `ggplot2`
umap_df <- data.frame(umap_data$layout) %>%
  # Turn samples_ids stored as rownames into column
  tibble::rownames_to_column("refinebio_accession_code") %>%
  # Bring only the variables that we want from the metadata into this data frame; match by sample ids
  dplyr::inner_join(dplyr::select(
    metadata, refinebio_accession_code,

>     histology, subgroup

Reviewer: Can you put this on the same line as above?

  ),
  by = "refinebio_accession_code"
  )
```

> 

Reviewer: Might add a chunk where we show a preview of this really quick to remind us of what variables we have ready for plotting before we get into it.

Now, let's plot the UMAP scores.

```{r}
# Make a scatterplot using `ggplot2` functionality
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2
  )
) +
  geom_point() + # This tells R that we want a scatterplot
  theme_classic() # This tells R to return a classic-looking plot with no gridlines

# Print out plot here
umap_plot
```

> 

Reviewer: Can we add a sentence that previews the next steps. Something like what we have elsewhere `It's hard to interpret our UMAP results without some metadata labels` https://github.com/AlexsLemonade/refinebio-examples/blob/7ab7fafbf3cbe171841cb4f5eac5184c123189a0/03-rnaseq/dimension-reduction_rnaseq_02_umap.Rmd#L344

> Now, let's label the data points based on their genotype subgroup since this is central to the subgroup specific based hypothesis in the [original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3683624/#SD1) [@Northcott2012].

Reviewer: This citation isn't coming out right.

```{r}
# Make a scatterplot with ggplot2
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2,
    color = subgroup # This will label points with different colors for each `subgroup`
  )
) +
  geom_point() +
  theme_classic()

# Print out plot here
umap_plot
```

We can add another label to our plot.
Let's also label the data points based on the histological subtype that each sample belongs to.

```{r}
# Make a scatterplot with ggplot2
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2,
    color = subgroup, # This will label points with different colors for each `subgroup`
    shape = histology # This will label points with different colors for each `histology` group
  )
) +
  geom_point() +
  theme_classic()

# Print out plot here
umap_plot
```

## Save annotated UMAP plot as a PNG

You can easily switch this to save to a JPEG or TIFF by changing the file name within the `ggsave()` function to the respective file suffix.

```{r}
# Save plot using `ggsave()` function
ggsave(file.path(
  plots_dir,
  "GSE37382_umap_scatterplot.png"

> ), # Replace with name relevant your plotted data

Reviewer: I suppose this comment should be moved up one. (I suspect `styler` might have moved this paranethesis down one line).

plot = umap_plot # Here we are giving the function the plot object that we want saved to file
)
```

# Resources for further learning

- [How UMAP works with visuals](https://umap-learn.readthedocs.io/en/latest/how_umap_works.html) [@McInnes2018-how-umap-works]
- [More on UMAP and its parameters](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) [@umap-vignette]

> - [Original paper on UMAP](https://arxiv.org/abs/1802.03426) by McInnes, Healy & Melville

Reviewer: ```suggestion
- [Original paper on UMAP](https://arxiv.org/abs/1802.03426) @McInnes2018
```

> [@McInnes2018]

Reviewer: ```suggestion
```

```

If the chunk above printed out `FALSE` to either of those tests, you won't be able to run this analysis _as is_ until those files are in the appropriate place.

If the concept of a "file path" is unfamiliar to you; we recommend taking a look at our [section about file paths](https://alexslemonade.github.io/refinebio-examples/01-getting-started/getting-started.html#an-important-note-about-file-paths-and-Rmds). 

# Using a different refine.bio dataset with this analysis?

If you'd like to adapt an example analysis to use a different dataset from [refine.bio](https://www.refine.bio/), we recommend placing the files in the `data/` directory you created and changing the filenames and paths in the notebook to match these files (we've put comments to signify where you would need to change the code).
We suggest saving plots and results to `plots/` and `results/` directories, respectively, as these are automatically created by the notebook.
From here you can customize this analysis example to fit your own scientific questions and preferences.

Note that UMAP does require more dimensions than PCA, though, so if you have a smaller 

> dataset (less than 20) and you obtain errors, you may not be able to run UMAP 

Reviewer: I might have missed this before, but this should all be on one line.

)) %>%
  # Tuck away the Gene id column as rownames
  tibble::column_to_rownames("Gene")
```

Let's ensure that the metadata and data are in the same sample order. 

```{r}
# Make the data in the order of the metadata
df <- df %>%
  dplyr::select(metadata$geo_accession)

# Check if this is in the same order
all.equal(colnames(df), metadata$geo_accession)
```

## Perform Uniform Manifold Approximation (UMAP)

Now we are going to use a combination of functions from the `umap` and the `ggplot2` packages to perform and visualize the results of the Uniform Manifold Approximation (UMAP) dimension reduction technique on our medulloblastoma samples.

In this code chunk, we are going to perform Uniform Manifold Approximation (UMAP) on our data and create a data frame using the UMAP scores and the variables from our metadata that we are going to use to annotate our plot later.

```{r}
# Perform Uniform Manifold Approximation (UMAP) using the `umap::umap()` function
umap_data <- umap::umap(
  t(df) # We have to transpose our data frame so we are obtaining UMAP scores for samples instead of genes
)

# Make into data frame for plotting with `ggplot2`

> umap_df <- data.frame(umap_data$layout) %>%

Reviewer: ```suggestion
umap_df <- data.frame(umap_data$layout) %>% # The umap values we need for plotting are stored in this `layout` element
```

    by = "refinebio_accession_code"
  )
```

Let's take a preview at the data frame we created in the chunk above.

```{r}
head(umap_df)
```

Now, let's plot the UMAP scores.

```{r}
# Make a scatterplot using `ggplot2` functionality
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2
  )
) +
  geom_point() + # This tells R that we want a scatterplot
  theme_classic() # This tells R to return a classic-looking plot with no gridlines

# Print out plot here
umap_plot
```

It's hard to interpret our UMAP results without some metadata labels on our plot.

Let's label the data points based on their genotype subgroup since this is central to the subgroup specific based hypothesis in the original paper [@Northcott2012].

```{r}
# Make a scatterplot with ggplot2
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2,
    color = subgroup # This will label points with different colors for each `subgroup`
  )
) +
  geom_point() +
  theme_classic()

# Print out plot here
umap_plot
```

It looks like Group 4 and SHH groups somewhat cluster with each other but Group 3 seems to also be clustering with Group 4.

> We can add another label to our plot to get more information on the clustering behavior of our data.

Reviewer: ```suggestion
We can add another label to our plot to potentially gain insight on the clustering behavior of our data.
```

Let's also label the data points based on the histological subtype that each sample belongs to.

```{r}
# Make a scatterplot with ggplot2
umap_plot <- ggplot(
  umap_df,
  aes(
    x = X1,
    y = X2,
    color = subgroup, # This will label points with different colors for each `subgroup`
    shape = histology # This will label points with different colors for each `histology` group
  )
) +
  geom_point() +
  theme_classic()

# Print out plot here
umap_plot
```

> Our histological subtype groups don't appear to be clustering in a discernible pattern but it can be good to try.

Reviewer: ```suggestion
Our histological subtype groups don't appear to be clustering in a discernible pattern. 
```

> cluster behavior for Group 3.

Reviewer: ```suggestion
```
